### ISSUE TITLE
Prioritize rules with long build times

### ISSUE THREAD

**Issue by knupfer** [ndmitchell/shake#312]
This makes it more probable to run until the end concurrently.  This is especially useful if the project is rather small and contains a rule which takes a lot of time and is independent of a lot of other rules.

**Comment by ndmitchell** [comment]
Agreed, that would be helpful, certainly for some workflows (I've been experiencing just such a workflow recently). There are two approaches to that - I can either require users to annotate rules as being most expensive, or drive it off the profile information. The former gives the user maximum control and works in the first run, the latter is more automatic. Which did you have in mind?

**Comment by knupfer** [comment]
For me the automatic one is much better.  Typically I build a lot of
times, therefore the first time isn't that important and it isn't always
clear which part will take very long when you compile the first time.

I like the idea, that shake learns from previous builds :)

**Comment by knupfer** [comment]
Perhaps it would be even better to prioritize rules which increase granularity, something like:
if rule A would allow rules A1, A2, A3, it is preferable over rule B which allows rule B1, B2 because it increases stronger the granularity.

This would also avoid some midway congestions in which all cores are waiting for a rule to terminate.

A sensible formula for calculating the granularity would be for example:
(n+1)*(1/t)
where n is the number of rules which would be workable after this rule and t is the average time for terminating this rule...

Would a pull request be appreciated to implement something like that?

**Comment by ndmitchell** [comment]
You might want to prioritise things which are "deeper", rather than "broader". For example, B1 might enable B4..B100, and so prioritising A would be the wrong thing to do. I guess the ideal metric would be a combination of how much time is blocked on this rule (e.g. the time to build this rule, plus the time of all dependents) and how much wall time is blocked on this rule (the blocked time reduced by the amount of parallelism). Stated alternatively, a blend of how long it takes to rebuild after touching this file with `-j1` and `-j0`.

However, I think just picking the slowest rule first would probably go a long way to removing the worst case behaviours, and then we could reevaluate. Alternatively, your formula might refine that a bit, it certainly sounds plausible.

I'd happily accept a pull request, although I wouldn't want this feature to slow down the general operation, which suggests using something cheap to compute. Information about how long a rule took last time is available nearby and could be plumbed through. Everything else would probably take more effort.

**Comment by knupfer** [comment]
My formula has the advantage in comparision to taking always the slowest that it can be applied without prior builds...

Perhaps it would be sensible to apply these priorities only in say 50% of the cases to avoid blocking on resources like IO.

I'll start to hack

**Comment by ndmitchell** [comment]
Yep, that is an advantage of your formula. And agreed, perhaps blend 50% this information and 50% randomness, and then you're unlikely to suffer even if you make bad guesses. I look forward to seeing what you come up with.
